pgx/src/datum/range.rs

```diff
@@ -0,0 +1,303 @@
+/*
+Portions Copyright 2019-2021 ZomboDB, LLC.
+Portions Copyright 2021-2022 Technology Concepts & Design, Inc. <support@tcdi.com>
+
+All rights reserved.
+
+Use of this source code is governed by the MIT license that can be found in the LICENSE file.
+*/
+
+use crate::{
+    pg_sys, void_mut_ptr, Date, FromDatum, IntoDatum, Numeric, Timestamp, TimestampWithTimeZone,
+};
+use pgx_pg_sys::{Oid, RangeBound};
+use std::default::Default;
+use std::marker::PhantomData;
+// use std::ops::Range;
+
+pub trait RangeSubType {
+    fn range_type_oid() -> Oid;
+}
+
+pub struct PgRange<T: FromDatum + IntoDatum + RangeSubType> {
+    ptr: *mut pg_sys::varlena,
+    range_type: *mut pg_sys::RangeType,
+    pub lower: RangeBound,
+    pub upper: RangeBound,
+    pub is_empty: bool,
+    __marker: PhantomData<T>,
+}
+
+impl<T: FromDatum + IntoDatum + RangeSubType> PgRange<T> {
+    unsafe fn from_pg(
+        ptr: *mut pg_sys::varlena,
+        range_type: *mut pg_sys::RangeType,
+        lower_bound: pg_sys::RangeBound,
+        upper_bound: pg_sys::RangeBound,
+        is_empty: bool,
+    ) -> Self {
+        PgRange {
+            ptr,
+            range_type,
+            lower: lower_bound,
+            upper: upper_bound,
+            is_empty,
+            __marker: PhantomData,
+        }
+    }
+
+    #[inline]
+    pub fn lower_val(&self) -> Option<T> {
+        if self.is_empty || self.lower.infinite {
+            None
+        } else {
+            unsafe { T::from_datum(self.lower.val.clone(), false) }
+        }
+    }
+
+    #[inline]
+    pub fn upper_val(&self) -> Option<T> {
+        if self.is_empty || self.upper.infinite {
+            None
+        } else {
+            unsafe { T::from_datum(self.upper.val.clone(), false) }
+        }
+    }
+
+    pub fn empty_range() -> Self {
+        Self::from_values_internal(None, None, true)
+    }
+
+    pub fn from_values(lower: Option<T>, upper: Option<T>) -> Self {
+        Self::from_values_internal(lower, upper, false)
+    }
+
+    pub fn from_bounds(lower_bound: RangeBound, upper_bound: RangeBound) {
+        unsafe {
+            Self::from_bounds_internal(lower_bound, upper_bound, false);
+        }
+    }
+
+    unsafe fn from_bounds_internal(
+        mut lower_bound: RangeBound,
+        mut upper_bound: RangeBound,
+        is_empty: bool,
+    ) -> Self {
+        let typecache =
+            pg_sys::lookup_type_cache(T::range_type_oid(), pg_sys::TYPECACHE_RANGE_INFO as i32);
+
+        let range_type =
+            pg_sys::make_range(typecache, &mut lower_bound, &mut upper_bound, is_empty);
+
+        Self::from_pg(
+            range_type as *mut pg_sys::varlena,
+            range_type,
+            lower_bound,
+            upper_bound,
+            is_empty,
+        )
+    }
+
+    fn from_values_internal(lower: Option<T>, upper: Option<T>, is_empty: bool) -> Self {
```

Comment:
In 0.5.0 we've started to walk towards more error/result support. You can observe the ` FromDatum::try_from_datum()`. Perhaps we could make this this `try_from_values_internal` and returning an `Result`? This would allow the `expect()`s to be removed.

Reply:
It would allow potentially replacing the `.expect("message")` with just a nice little `?`.

Reply:
good idea